gtsam/nonlinear/CustomFactor.h

```diff
@@ -0,0 +1,84 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    CustomFactor.h
+ * @brief   Class to enable arbitrary factors with runtime swappable error function.
+ * @author  Fan Jiang
+ */
+
+#pragma once
+
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+using namespace gtsam;
+
+namespace gtsam {
+
+typedef std::vector<Matrix> JacobianVector;
+
+class CustomFactor;
+
+typedef std::function<Vector(const CustomFactor&, const Values&, const JacobianVector*)> CustomErrorFunction;
```

Comment:
Reason for `JacobianVector *` to be const?

Reply:
It's a const pointer to an array (whose size is preallocated by GTSAM). 

Design note: pybind11 will invoke a copy if this is `JacobianVector &`, and modifications in Python will not be reflected.

I'll add this to the comments

Reply:
done